design/design.md

```diff
@@ -0,0 +1,251 @@
+Even if we use domain separation for $H$ and $H'$, we cannot guarantee that
+$G(x) = H(0, x) \oplus H'(1, x)$ is random looking.
+Define $H'(b, x) = H(\sim b, x)$; then $G(x) = 000...000$ for all $x$.
+
+## Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+## Mathematical rigor and production-readiness
+At this point, it is not our goal to make the implementation 100% correct and
+secure.
+We need to roll out a prototype so that our primary clients (Peras, etc) can
+incorporate it in their prototypes, and also so that the external community can
+start playing with it as well.
+This means that there might be tiny rounding errors in parameter derivation due
+to the use of floats, and there might be tiny errors when using a uniformly
+random hash output to sample some distribution needed for ALBA
+(e.g., Bernoulli, Uniform, Binomial, Poisson).
+We will also probably not rigorously analyze the $H_1$ optimization described in
+section 3.3 of the paper.
+
+## Parameter calculation
+ALBA has two kinds of parameters: user friendly parameters such as
+$\lambda_{sec}$, $\lambda_{rel}$, $n_p$, $n_f$,
+and algorithm specific parameters such as $u$, $d$ and $q$.
+
+Our implementation will have parameter derivation code converting user friendly
+to algorithm specific parameters.
+However, this should be decoupled from the ALBA algorithms:
+through documentation, we will encourage the developer to run parameter
+derivation manually in advance and then hardcode the output algorithm specific
+parameters in their application.
+The ALBA functions such as `prove` and `verify` will thus receive as input
+the algorithm specific parameters.
+
+The rational is that we want to be able to evolve parameter derivation
+(e.g., to optimize parameters for performance, or making the derivation logic
+more rigorous) while maintaining backward compatibility with previous versions
+of the library.
+
+## Deviations from the ALBA paper
+In the future, our implementation will likely deviate from the paper.
+
+One reason is that the definitions given there are quite poor.
+They are, unfortunately, inapplicable for most applications.
+For example, the assumption that there exists a weight oracle, reporting
+the weight of any data element given as input, is often unrealistic.
+In Mithril, for instance, light clients don't know the whole blockchain, but
+would need to verify that signatures in an ALBA certificate come from
+accounts with real stake.
+The ALBA paper simply assumes that the light clients can query the weight oracle
+to get the stake of the signature's creator, but in practice the ALBA
+certificate will need to contain for each signature the correct stake and
+a proof of its correctness (e.g., by supplying a merkle proof).
```

Comment:
Technically, we do not provide a proof of correctness in Peras/Leios, the verifiers has to cache the stake information in advance (or search it live).
You can say that the weight verification is not discussed in the paper but is necessary for security.